File: src/851.py
class Solution:
    def loudAndRich(self, richer: list[list[int]], quiet: list[int]) -> list[int]:
        n = len(quiet)

        graph = [[] for _ in range(n)]
        for a, b in richer:
            graph[b].append(a)  # b -> a: a богаче b

        memo = [-1] * n

        def dfs(person):
            if memo[person] != -1:
                return memo[person]

            # начинаем с самого себя
            quietest_person = person

            # смотрим на всех, кто богаче
            for richer_person in graph[person]:
                candidate = dfs(richer_person)  # самый тихий среди тех, кто богаче richer_person
                if quiet[candidate] < quiet[quietest_person]:
                    quietest_person = candidate

            memo[person] = quietest_person
            return quietest_person

        # запускаем dfs для каждого человека
        answer = []
        for i in range(n):
            answer.append(dfs(i))

        return answer
    

File: src/851.kt
class Solution {
    fun loudAndRich(richer: Array<IntArray>, quiet: IntArray): IntArray {
        val n = quiet.size
        val graph = Array(n) { mutableListOf<Int>() }
        for ((a, b) in richer) {
            graph[b].add(a)
        }
        val ans = IntArray(n) { -1 }
        fun dfs(u: Int): Int {
            if (ans[u] != -1) return ans[u]
            var best = u
            for (v in graph[u]) {
                val cand = dfs(v)
                if (quiet[cand] < quiet[best]) best = cand
            }
            ans[u] = best
            return best
        }
        for (i in 0 until n) dfs(i)
        return ans
    }
}
